Implement PDF build

rejahrehim opened the pull request:
This PR covers issue #234

- [x] This PR handles the issue and requires no additional PRs.
- [x] You have validated the need for this change.

**What did this PR accomplish?**

- Implement script to build PDF from markdown
- Upload the build to Github Release

rejahrehim commented:
Sample build command `npm run-script  pdfgen -- 4.0.1-alpha`
Install `md-to-pdf` (node package) before that `npm install -g md-to-pdf`

ThunderSon commented:
An issue still exist, as discussed on Slack, that the links are referencing localhost, just to keep track of it over here 😄

rejahrehim commented:
@ThunderSon Have a test, I have updated to work with the new directory structure. Fixed some issues with the inter page links. Many more to fix. Working on fixing those issues.

kingthorin commented:
Could we bundle the supporting files/directories into a single sub-dir (similar to what the www job does for it's ancillary files). Maybe just `/pdf/` in the root of the repo or something: Containing `scripts`, `assets`, and `pdf-config.json`.

kingthorin commented on scripts/make-pdf.sh:
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash
+# Script to create the PDF file for WSTG
+echo "Creating PDF for version $1"
```
I guess `$1` is coming from CLI position parsing. Could we reassign it to a meaningful name right away and use that going forward in the script, just to make reading/functionality more clear? Something like `TAG=$1` then use `$TAG` below (or `VERSION`/`$VERSION`).

This probably sounds trivial and over thought but I scrolled specifically to the end of the file to see something then had to read backward and into the workflow to find out what $1 actually was...

rejahrehim replied:
Sure.

rejahrehim commented:
> Could we bundle the supporting files/directories into a single sub-dir (similar to what the www job does for it's ancillary files). Maybe just `/pdf/` in the root of the repo or something: Containing `scripts`, `assets`, and `pdf-config.json`.

I will update that.

rejahrehim commented:
@kingthorin Need a thorough review.

kingthorin commented on .github/workflows/build-pdf.yml:
```diff
@@ -0,0 +1,52 @@
+on:
+  push:
+    tags: 
+      - '*'
```
Should we make this more specific?

I guess we'd need to establish a tagging standard, but we could probably come up with something quickly via slack.

kingthorin replied:
Maybe I'm overthinking it, I can't see anything else we'd use tags for....

kingthorin commented on .github/workflows/build-pdf.yml:
```diff
@@ -0,0 +1,52 @@
+    tags: 
+      - '*'
+jobs:
+  build:
```
Do we need to limit who can kick this process off?

I can't seem to find anything specific about who or what role is able to apply tags to a repo. Could do something like:
```yaml
    if: github.actor == 'kingthorin' || github.actor == 'ThunderSon' || github.actor == 'rejahrehim' || github.actor == 'victoriadrake'
```

rejahrehim replied:
Sure. I will update that.

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,155 @@
+find build/md -name "*README.md" | while IFS= read -r FILE; do mv -v "$FILE" "${FILE//README/0-0.0_README}"; done
+
+# Update version in pdf-config
+sed -i "s/{PDF Version}/$1/g" pdf/pdf-config.json
```
```suggestion
sed -i "s/{PDF Version}/$VERSION/g" pdf/pdf-config.json
```

rejahrehim replied:
Missed It.

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,155 @@
+# Create the final single markdown file.
+ls build/md | sort -n | while read x; do cat build/md/$x | sed -e 's/^# /<div style=\"page-break-after: always\;\"><\/div>\
+\
+# /' | sed 's/\[\([^\n]\+\)\]([^\n]\+.md#\([^\)]\+\)/[\1](#\2/' | sed 's/\(^#\{1\} \)\(\[\([0-9. ]*\)\(.*\)\]\(.*\)\(\?\:\n\+\|$\)\)/<h1><a href=\"#\4\">\4<\/a><\/h1>/'  | sed 's/\(^#\{2\} \)\(\[\([0-9. ]*\)\(.*\)\]\(.*\)\(\?\:\n\+\|$\)\)/<h2><a href=\"#\4\">\4<\/a><\/h2>/' | sed 's/\(^#\{3\} \)\(\[\([0-9. ]*\)\(.*\)\]\(.*\)\(\?\:\n\+\|$\)\)/<h3><a href=\"#\4\">\4<\/a><\/h3>/'| sed 's/\(^#\{4\} \ ...
```
This is basically white magic. With no description of the steps/requirements there's no way to review this monolithic stack of sed'ing. We kinda just have to trust you. (Same down around L44)

rejahrehim replied:
Let me try to explain it.

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,155 @@
+
+md-to-pdf  --config-file pdf/pdf-config.json  --pdf-options '{"margin":"0mm", "format": "A4"}' build/cover-$1.md
+
+md-to-pdf  --config-file pdf/pdf-config.json build/wstg-doc-$1.md
```
```suggestion
md-to-pdf  --config-file pdf/pdf-config.json build/wstg-doc-$VERSION.md
```

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,155 @@
+# Combine the files
+pdftk build/cover-$VERSION.pdf build/wstg-doc-$1.pdf cat output build/wstg-com-$VERSION.pdf
+
+
```
```suggestion
```

Remove a blank line.

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,155 @@
+pdftk build/cover-$VERSION.pdf build/wstg-doc-$1.pdf cat output build/wstg-com-$VERSION.pdf
+
+
+# Create capter PDF files for generating bookmarks
```
```suggestion
# Create chapter PDF files for generating bookmarks
```

kingthorin replied:
Would it make sense to create the chapters instead of the unified PDF above? Then combine the PDFs all into one?

rejahrehim replied:
Combining a PDF causes its size to be significantly larger.

https://apple.stackexchange.com/questions/8563/why-does-combining-a-pdf-cause-its-size-to-be-significantly-larger

https://stackoverflow.com/questions/19991144/pdftk-split-pdf-with-multiple-pages-but-total-size-grew?noredirect=1&lq=1

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,155 @@
+
+for f in build/pdf/*.md ; do md-to-pdf  --config-file pdf/pdf-config.json $f && rm $f; done
+
+# Generate capter details form individual chapter PDFfiles
```
```suggestion
# Generate chapter details form individual chapter PDF files
```

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,155 @@
+for f in build/pdf/*.pdf; do IFS='>>>' read -ra FILE <<< "$f"; for i in "${FILE[@]}"; do echo $i | sed 's/build\/pdf\/document/section: /'  | sed 's/\([0-9.]\+\)-\(.*\)\.pdf/\1 \; file: \2 \;/' | sed 's/Appx\.[A-Z]_\(.*\)/6 \; sectionTitle: Appendix \;  subsection: \1 \; sectionTitle: \1 \; subsection: /' | sed 's/\([0-9.]\+\)-\(.*\)/\1 \; sectionTitle: \2 \; subsection: /' | sed 's/_/ /g' | tr - ...
+
+# Generate Bookmark data from generated chapter details
+
```
```suggestion
```

ThunderSon commented:
@rejahrehim please pull latest in order to verify that the title changes made things better :smile

rejahrehim commented:
Blocked by #426 and #427

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,275 @@
+
+
+# In rare cases some PDF files has a blank page at the end due to newlines or box padding.
+# This affects the bookmark creation. So Remove those last pages if it is less than 12000 bytes
```
```suggestion
# This affects the bookmark creation. So Remove those last pages if it is less than 13000 bytes
```

kingthorin replied:
Per L145

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,275 @@
+    fi
+done;
+
+# Generate chapter details form individual chapter PDF files
```
```suggestion
# Generate chapter details from individual chapter PDF files
```

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -0,0 +1,275 @@
+
+# Create the single document Markdown file
+# Sed section 1: Add page break after each chapter
+# Sed section 2: Correct Markdown file links with fragment identifiers. Remove file path and keep fragmentidentifier alone.
```
```suggestion
# Sed section 2: Correct Markdown file links with fragment identifiers. Remove file path and keep fragment identifier alone.
```

kingthorin commented on pdf/scripts/make-pdf.sh:
```diff
@@ -81,7 +83,9 @@ python -c "import re; import sys; print(re.sub(r'href=\"(#[^\"]*)\"', lambda m:
 # Sed section 28: Move the number out of href
 sed 's/<h1 id=\"[0-9.]*-\(.*\)\">\(.*\)<\/h1>/<h1 id="\1">\2<\/h1>/' | \
-# Sed section 29: Add design to image name text
+# Sed section 29: Add design to image  and remove extra '\'
```
```suggestion
# Sed section 29: Add design to image and remove extra '\'
```

victoriadrake commented:
Close #234
